onnx/shape_inference/implementation.cc

```diff
@@ -486,21 +486,43 @@ class ShapeInferenceImplBase {
     }
   }
 
-  void process(const NodeProto& n, std::unordered_map<std::string, const AttributeProto*> attr_map) {
-    NodeProto copy_n(n);
-    // Add attribute information into the temporary node
-    copy_n.clear_attribute();
-    for (const auto& attr : n.attribute()) {
-      if (attr.has_ref_attr_name()) {
-        if (attr_map.count(attr.ref_attr_name())) {
-          auto copy_attr = *attr_map[attr.ref_attr_name()];
-          copy_attr.set_name(attr.name());
-          copy_n.add_attribute()->CopyFrom(copy_attr);
+  void replaceAttrParams(NodeProto& n, std::unordered_map<std::string, const AttributeProto*> attr_map) {
+    auto& attributes = *n.mutable_attribute();
+    for (auto attr_iter = attributes.begin(); attr_iter != attributes.end();) {
+      auto& attr = *attr_iter;
+      if (!attr.ref_attr_name().empty()) {
+        // Attribute-references must be replaced by the corresponding attribute-value in the call-node
+        // if the call-node contains the attribute. Otherwise, this attribute must be removed.
+        auto entry = attr_map.find(attr.ref_attr_name());
+        if (entry != attr_map.cend()) {
+          // Copy value of attribute, but retain original name:
+          std::string name = attr.name();
+          attr = *(entry->second);
+          attr.set_name(name);
+        } else {
+          attr_iter = attributes.erase(attr_iter);
+          continue;
         }
-      } else {
-        copy_n.add_attribute()->CopyFrom(attr);
       }
+      // Subgraphs must be recursively processed.
+      if (attr.has_g()) {
+        replaceAttrParams(*attr.mutable_g(), attr_map);
+      }
+      for (auto& graph : *attr.mutable_graphs())
+        replaceAttrParams(graph, attr_map);
```

Comment:
nit:
```suggestion
    for (auto& graph : *attr.mutable_graphs()) {
        replaceAttrParams(graph, attr_map);
    }
```

Reply:
re. the name Params: I find `replaceAttrValues` confusing. I could change to `replaceAttrRefs` I suppose. The point is that an Attribute inside a function-body could contain an attribute-value or it could be an attribute-parameter of the function itself. We want to replace the references to attribute-parameters by the values.